proposals/accepted/50-ide-test-integration.md

```diff
@@ -0,0 +1,219 @@
+regex searches, with support for hspec:
+ 1. [neo-test](https://github.com/mrcjkb/neotest-haskell)
+ 2. [vim-test](https://github.com/vim-test/vim-test)
+
+While these are valuable contributions to the ecosystem, their choice
+of regex for the collection of the user's test tree, and support for only
+specific frameworks are symptomatic of a lack of ecosystem coordination.
+
+## Motivation
+
+With the implementation of this proposal, the Haskell community will
+have gained a valuable IDE tool which will improve the user experience
+of Haskellers as well as make Haskellers more productive.
+
+Support for IDE integration of testing libraries will provide the following
+benefits:
+ 1. Ability for IDEs with Haskell support to provide inline hints for "Run
+ Tests" style commands.
+ 2. Ability for IDEs with Haskell support and support for a test explorer
+ view to display the user's test tree and run individual tests.
+ 3. A unified framework for testing libraries to provide integration with
+ IDEs. When a new testing framework is developed, they will be able to
+ provide the results in a Haskell datatype via serialization, which
+ avoids framework-specific test tree discovery logic.
+ 4. Potential to share test result rendering logic between different testing
+ frameworks, whereas currently each testing framework has their own test result
+ display logic.
+
+## Goals
+
+1. Create a new library, `hs-test`, with common datatypes for testing framework
+ test trees, and test results.
+2. Generate test trees and test results that can be serialized into JSON from
+Haskell's major testing frameworks, starting with `tasty`.
+3. Integrate test tree output output with VSC's Haskell extension, via VSC's [Testing API](https://code.visualstudio.com/api/extension-guides/testing).
+4. Stretch goal: support IDEs other than VSC.
+5. Stretch goal: support test results caching, and features such as "Rerun only failed tests".
+
+## Technical Proposal
+
+The `hs-test` library will contain the following data declarations:
+
+```haskell
+-- | Datatype representing all the tests that exist in a test
+-- suite that could be run.
+--
+-- `SrcLocs` are used to indicate where inline hints should be displayed
+-- in an IDE.
+data TestTree =
+  TestGroup {name :: Text, subTrees :: [TestTree], location :: SrcLoc}
+  | TestCase {name :: Text, location :: SrcLoc}
+
+data RunTestsRequest =
+  RunTestsRequest {path :: NonEmpty Text}
+
+-- | A lazy list in order to support arbitrary execution
+-- order and streaming between the test runner and IDE
+data TestResults = TestResults [TestCaseResult]
+
+data TestCaseResult = TestCaseResult {
+    -- | Represents the path to get to this test case from the
+    -- TestTree, where each element represents a node in the original
+    -- TestTree.
+    path :: NonEmpty Text,
+    result :: TestResult,
+    timeTaken :: Maybe Double
+  }
+
+data TestResult = Success | Skipped | Failure FailureReason
```

Comment:
Having timing information is great. Is it enough?

Using Haskell datatypes as a means of thought is great, but I'd almost prefer the proposal to just define the JSON format. This would allow subtypes to be built (read: with extra fields) to allow the community to experiment with what additional information is useful on top of a common core. For instance, QuickCheck tests might want to show how many tests passed vs were skipped, and that would be useful feedback in my editor.